### PR TITLE
feat : add log by step

### PR DESCRIPTION
**Pull request by sguiheux**

### CODE REVIEW

**Comment by fsamin** [comment]
Plugin logs should go though worker http server and managed by the worker goroutine

File: cli/cds/pipeline/log.go [hunk 1]
```diff
@@ -57,21 +57,26 @@ func showBuildPipeline(cmd *cobra.Command, args []string) {
...
+					v,
+				)
+				w.Flush()
+			}
 
-		// Exit 1 if pipeline fail
-		if l.ID == 0 && strings.Contains(l.Value, "status: Fail") {
-			sdk.Exit("")
+			// Exit 1 if pipeline fail
+			if l.ID == 0 && strings.Contains(l.Value, "status: Fail") {
```
Comment (yesnault) [inline cli/cds/pipeline/log.go]: const ? reused in run.go

File: engine/api/build.go [hunk 2]
```diff
@@ -23,6 +23,59 @@ import (
 	"github.com/ovh/cds/sdk"
 )
 
+func updateStepStatusHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c *context.Ctx) error {
```
Comment (fsamin) [inline engine/api/build.go]: Tests ?

File: engine/api/build.go [hunk 3]
```diff
@@ -23,6 +23,59 @@ import (
...
+	vars := mux.Vars(r)
+	buildIDString := vars["id"]
+
+	buildID, errID := strconv.ParseInt(buildIDString, 10, 64)
+	if errID != nil {
+		log.Warning("updateStepStatusHandler> buildID must be an integer: %s\n", errID)
+		return sdk.ErrInvalidID
```
Comment (fsamin) [inline engine/api/build.go]: sdk.ErrWrongRequest

File: engine/worker/run.go [hunk 4]
```diff
@@ -210,32 +272,52 @@ func logger(inputChan chan sdk.Log) {
...
 				l.Value = strings.Trim(strings.Replace(l.Value, "\n", " ", -1), " \t\n") + "\n"
-				logs = append(logs, l)
-			}
 
-			if len(logs) == 0 {
-				continue
+				// First log
+				if currentStepLog == nil {
+					currentStepLog = &l
+				} else {
```
Comment (yesnault) [inline engine/worker/run.go]: else if

File: engine/worker/run.go [hunk 5]
```diff
@@ -210,32 +272,52 @@ func logger(inputChan chan sdk.Log) {
...
+			for _, l := range logs {
+				// Buffer log list is empty, sending batch to API
+				data, err := json.Marshal(l)
+				if err != nil {
+					fmt.Printf("Error: cannot marshal logs: %s\n", err)
+					continue
+				}
+
+				path := fmt.Sprintf("/build/%d/log", l.PipelineBuildJobID)
+				_, _, err = sdk.Request("POST", path, data)
```
Comment (yesnault) [inline engine/worker/run.go]: oneline

File: sdk/build.go [hunk 6]
```diff
@@ -21,11 +21,23 @@ type PipelineBuildJob struct {
 	PipelineBuildID int64       `json:"pipeline_build_id,omitempty" db:"pipeline_build_id"`
 }
 
+type ExecutedJob struct {
```
Comment (yesnault) [inline sdk/build.go]: doc?

File: sdk/build.go [hunk 7]
```diff
@@ -21,11 +21,23 @@ type PipelineBuildJob struct {
...
+	Job
+	StepStatus []StepStatus `json:"step_status" db:"-"`
+	Reason     string
+}
+
+type StepStatus struct {
```
Comment (yesnault) [inline sdk/build.go]: status not enum ? doc ?

File: sdk/pipeline.go [hunk 8]
```diff
@@ -430,47 +408,62 @@ func StreamPipelineBuild(key, appName, pipelineName, env string, buildID int, fo
...
-										//Get log from the trigger's channel and push it to the main channel
-										for l := range triggerCh {
-											ch <- l
-										}
-										wg.Done()
-									}(ch, triggerCh)
+				for i := range logs {
+					if logs[i].ID != 0 {
+						ch <- logs[i]
+					}
```
Comment (yesnault) [inline sdk/pipeline.go]: continue, remove if below

File: engine/api/build.go [hunk 9]
```diff
@@ -23,6 +23,59 @@ import (
...
+		}
+	}
+	if !found {
+		pbJob.Job.StepStatus = append(pbJob.Job.StepStatus, step)
+	}
+
+	var errmarshal error
+	pbJob.JobJSON, errmarshal = json.Marshal(pbJob.Job)
+	if errmarshal != nil {
+		log.Warning("updateStepStatusHandler> Cannot marshall job: %s\n", errR)
```
Comment (fsamin) [inline engine/api/build.go]: errmarshal

File: engine/api/build.go [hunk 10]
```diff
@@ -23,6 +23,59 @@ import (
...
+		return errmarshal
+	}
+	if err := pipeline.UpdatePipelineBuildJob(db, pbJob); err != nil {
+		log.Warning("updateStepStatusHandler> Cannot update pipeline build job: %s\n", errR)
```
Comment (fsamin) [inline engine/api/build.go]: err

File: engine/api/build.go [hunk 11]
```diff
@@ -23,6 +23,59 @@ import (
...
+
+	// Get body
+	data, errR := ioutil.ReadAll(r.Body)
+	if errR != nil {
+		log.Warning("updateStepStatusHandler> Cannot read body: %s\n", errR)
+		return sdk.ErrWrongRequest
+	}
+	var step sdk.StepStatus
+	if err := json.Unmarshal(data, &step); err != nil {
+		log.Warning("updateStepStatusHandler> Cannot unmarshall body: %s\n", errR)
```
Comment (yesnault) [inline engine/api/build.go]: errR --> err

File: engine/api/build_log.go [hunk 12]
```diff
@@ -19,33 +20,107 @@ import (
...
 	buildNumberS := vars["build"]
 	appName := vars["permApplicationName"]
+	pipelineActionIDString := vars["actionID"]
+	stepOrderString := vars["stepOrder"]
 
-	// Get offset
-	err := r.ParseForm()
+	stepOrder, errInt := strconv.Atoi(stepOrderString)
+	if errInt != nil {
+		return sdk.ErrInvalidID
```
Comment (fsamin) [inline engine/api/build_log.go]: sdl.ErrWrongRequest

File: engine/api/build_log.go [hunk 13]
```diff
@@ -19,33 +20,107 @@ import (
...
+	var env *sdk.Environment
+	envName := r.FormValue("envName")
+	if envName == "" || envName == sdk.DefaultEnv.Name {
+		env = &sdk.DefaultEnv
+	} else {
+		var errEnv error
+		env, errEnv = environment.LoadEnvironmentByName(db, projectKey, envName)
+		if errEnv != nil {
+			log.Warning("getStepBuildLogsHandler> Cannot load environment %s: %s\n", envName, errEnv)
+			return sdk.ErrUnknownEnv
```
Comment (fsamin) [inline engine/api/build_log.go]: return errEnv

File: engine/api/build_log.go [hunk 14]
```diff
@@ -19,33 +20,107 @@ import (
...
-		return err
+		log.Warning("getStepBuildLogsHandler> Cannot load pipeline %s: %s\n", pipelineName, err)
+		return sdk.ErrPipelineNotFound
+	}
 
+	// Check that application exists
+	a, err := application.LoadApplicationByName(db, projectKey, appName)
+	if err != nil {
+		log.Warning("getStepBuildLogsHandler> Cannot load application %s: %s\n", appName, err)
+		return sdk.ErrApplicationNotFound
```
Comment (fsamin) [inline engine/api/build_log.go]: return err

File: engine/api/build_log.go [hunk 15]
```diff
@@ -19,33 +20,107 @@ import (
...
 	}
-	offsetS := r.FormValue("offset")
-	var offset int64
-	if offsetS != "" {
-		offset, err = strconv.ParseInt(offsetS, 10, 64)
+
+	// if buildNumber is 'last' fetch last build number
+	var buildNumber int64
+	if buildNumberS == "last" {
+		bn, err := pipeline.GetLastBuildNumberInTx(db, p.ID, a.ID, env.ID)
```
Comment (fsamin) [inline engine/api/build_log.go]: I think that GetLastBuildNumberInTx is useless since we are using interface sqlexecutor

Comment (fsamin) [inline engine/api/build_log.go]: nop, start a `tx` in the handler; use `GetLastBuildNumberIn(...)` and remove the function `GetLastBuildNumberInTx`

File: engine/api/build_log.go [hunk 16]
```diff
@@ -139,41 +213,21 @@ func getActionBuildLogsHandler(w http.ResponseWriter, r *http.Request, db *gorp.
...
-		log.Warning("getActionBuildLogsHandler> actionID should be an integer : %s\n", err)
-		return err
-
-	}
-
-	// Get offset
-	if err := r.ParseForm(); err != nil {
-		log.Warning("getActionBuildLogsHandler> cannot parse form: %s\n", err)
+		log.Warning("getPipelineBuildJobLogsHandler> actionID should be an integer : %s\n", err)
 		return err
```
Comment (fsamin) [inline engine/api/build_log.go]: sdk.ErrWrongRequest

File: engine/api/pipeline/awol.go [hunk 17]
```diff
@@ -72,18 +72,18 @@ func killAWOLAction(db *gorp.DbMap, pbJobData awolPipelineBuildJob) error {
 }
 
 // SELECT action_build.id
```
Comment (fsamin) [inline engine/api/pipeline/awol.go]: why this sql query in comment ?

File: engine/api/pipeline/build_log.go [hunk 18]
```diff
@@ -1,59 +1,68 @@
...
-
-	if start > 0 {
-		query = fmt.Sprintf("%s AND id > %d", query, start)
+// InsertLog insert build log into database
+func InsertLog(db gorp.SqlExecutor, l *sdk.Log) error {
+	dbmodel := database.Log(*l)
+	if err := db.Insert(&dbmodel); err != nil {
+		return err
 	}
+	return nil
```
Comment (fsamin) [inline engine/api/pipeline/build_log.go]: i miss `*l = *dbmodel` or something like that

File: engine/api/build_log.go (hunk 15 above)
Comment (sguiheux) [inline engine/api/build_log.go]: We will refactor CLI after and it will not send  last anymore

File: engine/api/build.go (hunk 3 above)
Comment (sguiheux) [inline engine/api/build.go]: ErrInvalidID for non integer in URL

**Pull request merged**